### PR TITLE
improve the config-gateway function

### PR DESCRIPTION
**Pull request by LeZhang2016**
- add a property named "Thread:OnMeshPrefixes" which indicates the information about gateway (prefix, flags, and priority).
- add "priority" for configured prefix.
- provide "flags" (preferred, slaac, onMesh) for OT border router setting

### CODE REVIEW

**Comment by darconeous** [comment]
I think this commit is missing some files you added.

**Comment by LeZhang2016** [comment]
@darconeous, Thanks a lot for your reminder. I have fixed it.

File: src/ipc-dbus/DBusIPCAPI_v1.cpp [hunk 1]
```diff
@@ -1028,19 +1028,28 @@ DBusIPCAPI_v1::interface_config_gateway_handler(
...
+	int16_t priority_raw;
+	NCPControlInterface::OnMeshPrefixPriority priority(NCPControlInterface::PREFIX_MEDIUM_PREFERENCE);
 
 	did_succeed = dbus_message_get_args(
 		message, NULL,
 		DBUS_TYPE_BOOLEAN, &default_route,
 		DBUS_TYPE_ARRAY, DBUS_TYPE_BYTE, &prefix, &prefix_len_in_bytes,
 		DBUS_TYPE_UINT32, &preferred_lifetime,
 		DBUS_TYPE_UINT32, &valid_lifetime,
+		DBUS_TYPE_BOOLEAN, &preferred,
```
Comment (abtink) [inline src/ipc-dbus/DBusIPCAPI_v1.cpp]: This changes the dbus message format and  breaks compatibility with sw using existing version (using existing dbus message format).

We should check for the case where the new parameters are not provided, which should cause a `did_succeed` failure, and then parse it up to old parameters (use default values for new ones).  `DBusIPCAPI_v1::interface_route_add_handler()` is an example of how it's done for another command.

File: src/wpantund/NCPInstanceBase.cpp [hunk 2]
```diff
@@ -435,6 +437,19 @@ NCPInstanceBase::get_property(
 	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_NetworkNodeType)) {
 		cb(0, boost::any(node_type_to_string(mNodeType)));
 
+	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_ThreadOnMeshPrefixes)) {
+		std::list<std::string> result;
+		std::map<struct in6_addr, GlobalAddressEntry>::const_iterator it;
+		static const char flag_lookup[] = "ppPSDCRM";
+		char address_string[INET6_ADDRSTRLEN];
+		for ( it = mOnMeshPrefixes.begin();
+			  it != mOnMeshPrefixes.end();
```
Comment (abtink) [inline src/wpantund/NCPInstanceBase.cpp]: nit: The alignment/spacing for this `for`.

File: src/ipc-dbus/DBusIPCAPI_v1.cpp (hunk 1 above)
Comment (LeZhang2016) [inline src/ipc-dbus/DBusIPCAPI_v1.cpp]: @abtink, Thanks a lot for your comments, I will modify this part according to your suggestion.

File: src/wpantund/NCPInstanceBase.cpp (hunk 2 above)
Comment (LeZhang2016) [inline src/wpantund/NCPInstanceBase.cpp]: Thanks, I will modify it.

**Comment by abtink** [comment]
Looks great. Thanks for doing this.

**Pull request merged**